### ISSUE TITLE
Add overload to Path.GetFullPath() to specify base path

### ISSUE THREAD

**Issue by JeremyKuhne** [dotnet/corefx#25535]
We should add an overload that allows specifying the base path to resolve against if the given `path` isn't fully qualified.

## Rationale and Usage

Resolving partially qualified paths (paths that are relative to the current directory) is error prone as it depends on the process-wide state of the current working directory. Allowing users to specify a virtual "working directory" will avoid many common errors caused by setting the current directory to resolve paths "correctly".

``` C#
foreach (string file in sourceFiles)
{
    resolvedSources.Add(Path.GetFullPath(file, projectRoot));
}
```

## Proposed API

``` C#
namespace System.IO
{
    public static class Path
    {
        public static string GetFullPath(string path, string basePath);
    }
}
```

## Implementation Notes

- `basePath` will throw ArgumentNullException if null.
- `basePath` will throw ArgumentException if it does not meet `Path.IsPathFullyQualified()`. 
- If `path` does not meet `Path.IsPathFullyQualified()`, it will be combined with the `basePath` before resolving.
- If the given path is drive relative on Windows (`C:foo`) we will combine with the `basePath` **if** the drive matches. If it does not, we will assume it is meant to be at the root of the specified drive. (Option: we could throw in this case...)

**Comment by jkotas** [comment]
Should the arguments be in the opposite order to match `Path.Combine(string path1, string path2)` ?

I would expect that this API is often going to be used as a more correct replacement of `Path.Combine`. It may be nice to have the arguments in the same order.

**Comment by JeremyKuhne** [comment]
> Should the arguments be in the opposite order to match Path.Combine(string path1, string path2) ?

It would make more logical sense, and I would actually prefer `(basePath, path)`. I thought, however, that we were trying to match existing order for arguments so I started the other way around.

**Comment by karelz** [comment]
FYI: The API review discussion was recorded - see https://youtu.be/9F_NsviGT0Y?t=18 (9 min duration)

**Comment by JeremyKuhne** [comment]
API review wanted to match the initial proposed order (path, basePath). This one needs good test coverage, particularly on Windows specific paths:

- Base device paths: `\\?\` & `\\.\`
    - these should **only** be recognized if they are in their canonical format, we have existing helpers for identifying
    - these are **only** valid for the base path, if the `path` is in this format it must, by definition, be fully qualified
    - the result will need to be passed through RemoveRelativeSegments(), which is currently only in the Unix impl: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/IO/Path.Unix.cs
    - do not normalize separators and do not consider AltDirectorySeparator `/` to be a directory separator when the base is in this format
- Path is current drive rooted: `\`
    - need to add to the root of the base path- want to make sure to check device paths for this case as well
- Path is specific drive rooted: `C:foo`
    - *only* consider this case if in the correct drive format (there is a helper- `IsValidDriveChar()`, followed by `:` and **no** directory separator)
    - if the `basePath` drive matches `path`, combine the two, otherwise just insert `\` after the `:`

Other notes in the issue description should be followed as well, of course.

**Comment by danmosemsft** [comment]
5 days assuming it's someone less familiar with the area.

**Comment by Anipik** [comment]
Can somebody give an example of how api should work in case of device paths (1  & 2 case)

**Comment by JeremyKuhne** [comment]
> Can somebody give an example of how api should work in case of device paths (1 & 2 case)

@Anipik My comments above describe what to do. `PathInternal.IsDevice()` will tell you if a path is a device path. So, given that, you'd be doing something roughly like this:

``` C#
if (Path.IsFullyQualified(path))
{
    // basePath is meaningless, call normal Path.GetFullPath()
    return Path.GetFullPath(path);
}

// Windows ONLY:

// Look for C:, etc. with no slash
if (path.Length > 1 && path[1] == ':' && PathInternal.IsValidDriveChar(path[0]))
{
    // Shouldn't have a slash after the colon- would have been fully qualified
    Debug.Assert(path.Length == 2 || !PathInternal.IsDirectorySeparator(path[3]));
    
    // Return as described in https://github.com/dotnet/corefx/issues/25535#issuecomment-348363953
}

if (PathInternal.IsDevice(basePath))
{
    // Combine as described in https://github.com/dotnet/corefx/issues/25535#issuecomment-348363953
    // https://github.com/dotnet/coreclr/blob/688b75c143aa0e080f386a04c74b13b3fc9877bf/src/mscorlib/shared/System/IO/Path.Unix.cs#L51
    // Don't want to eat segments above the root, make sure you have tests! ("C:\.." should come out to "C:\", or "\\?\C:\.." should result in "\\?\C:\", etc.)
    return Path.RemoveRelativeSegments(PathHelpers.CombineNoChecks(basePath, path), PathInternal.GetRootLength(basePath));
}

// "Normal" path. Combine and resolve (same for Windows/Unix)
return (Path.GetFullPath(PathHelpers.CombineNoChecks(basePath, path)));
```

**Comment by JeremyKuhne** [comment]
@Anipik please close this out once you think you've fully implemented and resolved all outstanding issues.

**Comment by Anipik** [comment]
yeah I am just waiting on the new version of coreclr in corefx
